File: app/src/main/java/com/innov8/memeit/Adapters/CommentsAdapter.java
package com.innov8.memeit.Adapters;

import android.content.Context;
import android.content.Intent;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.afollestad.materialdialogs.DialogAction;
import com.afollestad.materialdialogs.MaterialDialog;
import com.innov8.memegenerator.adapters.ListAdapter;
import com.innov8.memegenerator.adapters.MyViewHolder;
import com.innov8.memeit.Activities.ProfileActivity;
import com.innov8.memeit.CustomViews.ProfileDraweeView;
import com.innov8.memeit.KUtilsKt;
import com.innov8.memeit.R;
import com.memeit.backend.MemeItMemes;
import com.memeit.backend.MemeItUsers;
import com.memeit.backend.dataclasses.Comment;
import com.memeit.backend.utilis.OnCompleteListener;

import org.jetbrains.annotations.NotNull;

import java.security.InvalidParameterException;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.Group;
import okhttp3.ResponseBody;

/**
 * Created by Jv on 7/5/2018.
 */

public class CommentsAdapter extends ListAdapter<Comment> {

    boolean isPostingComment;

    float size;
    public static final int LIKE = 0;
    public static final int DISLIKE = 1;
    public static final int ADD = 0;
    public static final int REMOVE = 1;

    public CommentsAdapter(@NotNull Context mContext) {
        super(mContext, R.layout.list_item_comment_new);
        size = mContext.getResources().getDimension(R.dimen.profile_mini_size);
    }

    @NotNull
    @Override
    public MyViewHolder<Comment> createViewHolder(@NotNull View view) {
        return new CommentViewHolder(view);
    }

    protected class CommentViewHolder extends MyViewHolder<Comment> implements View.OnClickListener {
        private final TextView commentV;
        private final TextView dateV;

        private final ProfileDraweeView posterPicV;
        private final TextView posterNameV;
        private ImageView edit;
        private ImageView delete;
        private TextView like;
        private TextView dislike;
        Group ownCommentViews;


        public CommentViewHolder(View itemView) {
            super(itemView);
            commentV = itemView.findViewById(R.id.list_comment);
            dateV = itemView.findViewById(R.id.list_item_date);
            posterPicV = itemView.findViewById(R.id.comment_poster_pp);
            posterNameV = itemView.findViewById(R.id.list_name);
            edit = itemView.findViewById(R.id.edit);
            delete = itemView.findViewById(R.id.delete);
            like = itemView.findViewById(R.id.like_comment);
            dislike = itemView.findViewById(R.id.dislike_comment);
            ownCommentViews = itemView.findViewById(R.id.own_comment);
            posterPicV.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Intent i = new Intent(getMContext(), ProfileActivity.class);
                    i.putExtra("user", getItemAt(getItem_position()).getPoster().toUser());
                    getMContext().startActivity(i);
                }
            });
            delete.setOnClickListener(this);
            edit.setOnClickListener(this);
            like.setOnClickListener(this);
            dislike.setOnClickListener(this);
        }


        public void bind(final Comment mComment) {
            KUtilsKt.loadImage(posterPicV,mComment.getPoster().getProfileUrl(),size,size);
            commentV.setText(mComment.getComment());
            posterPicV.setText(KUtilsKt.prefix(mComment.getPoster().getName()));
            posterNameV.setText(mComment.getPoster().getName());
            dateV.setText(KUtilsKt.formatDate(mComment.getDate()));
            like.setText(String.valueOf(mComment.getLikeCount()));
            dislike.setText(String.valueOf(mComment.getDislikeCount()));
            boolean isOwnComment = MemeItUsers.getInstance().getMyUser(getMContext()).getUserID().equals(mComment.getPosterID());
            if (isOwnComment)
                ownCommentViews.setVisibility(View.GONE);
            else
                ownCommentViews.setVisibility(View.VISIBLE);
        }

        @Override
        public void onClick(View v) {
            final Comment comment = getItemAt(getItem_position());
            switch (v.getId()) {
                case R.id.like_comment:
                    if (comment.isLikedByMe()) {
                        MemeItMemes.getInstance().removeLikeComment(comment.getCommentID(), generateOnComplete(LIKE,REMOVE,getItem_position()));
                    } else {
                        MemeItMemes.getInstance().likeComment(comment.getCommentID(), generateOnComplete(LIKE,ADD,getItem_position()));
                    }
                    break;
                case R.id.dislike_comment:
                    if (comment.isDislikedByMe()) {
                        MemeItMemes.getInstance().removeDislikeComment(comment.getCommentID(), generateOnComplete(DISLIKE,REMOVE,getItem_position()));
                    } else {
                        MemeItMemes.getInstance().dislikeComment(comment.getCommentID(), generateOnComplete(DISLIKE,ADD,getItem_position()));
                    }
                    break;
                case R.id.delete:
                    new MaterialDialog.Builder(getMContext())
                            .title("Delete comment?")
                            .positiveText("Yes")
                            .negativeText("No")
                            .onPositive(new MaterialDialog.SingleButtonCallback() {
                                @Override
                                public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
                                    MemeItMemes.getInstance().deleteComment(comment.getMemeID(), comment.getCommentID(), new OnCompleteListener<ResponseBody>() {
                                        @Override
                                        public void onSuccess(ResponseBody responseBody) {
                                            Toast.makeText(getMContext(), "Deleted.", Toast.LENGTH_SHORT).show();
                                            remove(comment);
                                        }

                                        @Override
                                        public void onFailure(Error error) {
                                            Toast.makeText(getMContext(), "An error has occured. Please try again.", Toast.LENGTH_SHORT).show();
                                            Log.w("Deleting  comment", error.getMessage());
                                        }
                                    });
                                }
                            }).show();
                    break;
                case R.id.edit:

                    new MaterialDialog.Builder(getMContext())
                            .title("Edit comment")
                            .input("Comment", comment.getComment(), false, new MaterialDialog.InputCallback() {
                                @Override
                                public void onInput(@NonNull MaterialDialog dialog, final CharSequence input) {
                                    if (TextUtils.isEmpty(input))
                                        return;
                                    MemeItMemes.getInstance().editComment(comment.getCommentID(), input.toString(), new OnCompleteListener<ResponseBody>() {
                                        @Override
                                        public void onSuccess(ResponseBody o) {
                                            Toast.makeText(getMContext(), "Edited.", Toast.LENGTH_SHORT).show();
                                            comment.setComment(input.toString());
                                            commentV.setText(input.toString());
                                        }

                                        @Override
                                        public void onFailure(Error error) {
                                            Toast.makeText(getMContext(), "Editing has failed because of a problem. Please try again.", Toast.LENGTH_SHORT).show();
                                        }
                                    });
                                }
                            }).show();
                    break;

            }
        }
    }
    public OnCompleteListener generateOnComplete(final int type, final int addOrRemove, final int position){
        return new OnCompleteListener() {
            @Override
            public void onSuccess(Object o) {
                Comment comment = getItemAt(position);
                if(type == LIKE){
                    comment.setLikedByMe(addOrRemove == ADD);
                    comment.setLikeCount(addOrRemove == ADD
                            ? comment.getLikeCount() + 1
                            : comment.getLikeCount() - 1);
                    comment.setDislikeCount(addOrRemove == ADD
                            ? comment.getDislikeCount() - 1
                            : comment.getDislikeCount());
                }
                else if(type == DISLIKE){
                    comment.setDislikedByMe(addOrRemove == ADD);
                    comment.setDislikeCount(addOrRemove == ADD
                            ? comment.getDislikeCount() + 1
                            : comment.getDislikeCount() - 1);
                    comment.setLikeCount(addOrRemove == ADD
                            ? comment.getLikeCount() - 1
                            : comment.getLikeCount());
                }
                else throw new InvalidParameterException("Parameter \"Type\" must be either LIKE or DISLIKE");
                notifyItemChanged(position);
            }

            @Override
            public void onFailure(Error error) {
                Toast.makeText(getMContext(),"An error has occured",Toast.LENGTH_SHORT).show();
            }
        };
    }
}

File: app/src/main/java/com/innov8/memeit/adapters/CommentsAdapter.kt
package com.innov8.memeit.adapters

import android.content.Context
import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.Typeface
import android.text.Spannable
import android.text.TextUtils
import android.text.style.RelativeSizeSpan
import android.text.style.StyleSpan
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.core.text.set
import androidx.core.text.toSpannable
import com.afollestad.materialdialogs.MaterialDialog
import com.github.ybq.android.spinkit.style.CubeGrid
import com.innov8.memegenerator.adapters.MyViewHolder
import com.innov8.memeit.R
import com.innov8.memeit.activities.ProfileActivity
import com.innov8.memeit.commons.min
import com.innov8.memeit.commons.views.MemeItTextView
import com.innov8.memeit.commons.views.ProfileDraweeView
import com.innov8.memeit.utils.*
import com.memeit.backend.MemeItClient
import com.memeit.backend.MemeItMemes
import com.memeit.backend.call
import com.memeit.backend.models.Comment
import kotlinx.android.synthetic.main.list_item_meme.view.*

/**
 * Created by Jv on 7/5/2018.
 */

class CommentsAdapter(context: Context) : SimpleELEListAdapter<Comment>(context, R.layout.list_item_comment) {
    override fun createViewHolder(view: View): MyViewHolder<Comment> {
        return CommentViewHolder(view)
    }


    override var emptyDrawableId: Int = R.drawable.no_comments
    override var errorDrawableId: Int = R.drawable.ic_no_internet
    override var emptyDescription: String = "No comments yet."
    override var errorDescription: String = "Couldn't load comments"
    override var emptyActionText: String? = ""
    override var errorActionText: String? = "Try Again"
    override val loadingDrawable = CubeGrid().apply {
        color = Color.rgb(255, 100, 0)
    }

    internal val linkActions by lazy {
        generateTextLinkActions(context)
    }

    inner class CommentViewHolder(itemView: View) : MyViewHolder<Comment>(itemView), View.OnClickListener {
        private val commentV: MemeItTextView = itemView.findViewById(R.id.list_comment)
        private val dateV: TextView = itemView.findViewById(R.id.list_item_date)
        private val posterPicV: ProfileDraweeView = itemView.findViewById(R.id.comment_poster_pp)
        private val edit: ImageView = itemView.findViewById(R.id.edit)
        private val delete: ImageView = itemView.findViewById(R.id.delete)
        private val likeCount: TextView = itemView.findViewById(R.id.like_comment)
        private val dislikeCount: TextView = itemView.findViewById(R.id.dislike_comment)
        private val like: ImageView = itemView.findViewById(R.id.like)
        private val dislike: ImageView = itemView.findViewById(R.id.dislike)


        init {
            posterPicV.setOnClickListener {
                val i = Intent(context, ProfileActivity::class.java)
                i.putExtra("user", getItemAt(item_position).poster)
                context.startActivity(i)
            }
            delete.setOnClickListener(this)
            edit.setOnClickListener(this)
            likeCount.setOnClickListener(this)
            dislikeCount.setOnClickListener(this)
            like.setOnClickListener(this)
            dislike.setOnClickListener(this)
            commentV.onLinkClicked = linkActions
        }

        private fun applySpan(text: String, vararg words: String): Spannable =
                text.toSpannable().apply {
                    words.forEach {
                        val i = text.indexOf(it)
                        this[i..i + it.length] = StyleSpan(Typeface.BOLD)
                        this[i..i + it.length] = RelativeSizeSpan(1.05f)
                    }
                }


        override fun bind(t: Comment) {
            posterPicV.loadImage(t.poster!!.imageUrl)
            commentV.text = applySpan("${t.poster!!.name} ${t.comment}", t.poster!!.name!!)
            posterPicV.setText(t.poster!!.name.prefix())
            dateV.text = t.date!!.formateAsDate()
            likeCount.text = t.likeCount.toString()
            dislikeCount.text = t.dislikeCount.toString()
            val isOwnComment = MemeItClient.myUser!!.id == t.poster!!.uid
            if (isOwnComment) {
                edit.visibility = View.VISIBLE
                delete.visibility = View.VISIBLE
            } else {
                edit.visibility = View.GONE
                delete.visibility = View.GONE
            }
            like.setColorFilter(if (t.isLikedByMe) R.color.colorAccent.color(context) else Color.parseColor("#999999"))
            dislike.setColorFilter(if (t.isDislikedByMe) R.color.colorAccent.color(context) else Color.parseColor("#999999"))
        }

        override fun onClick(v: View) {
            val comment = getItemAt(item_position)
            val pos = item_position
            when (v.id) {
                R.id.like_comment, R.id.like -> {
                    if (comment.isLikedByMe) {
                        MemeItMemes.removeLikeComment(comment.id!!).call {
                            comment.likeCount = (comment.likeCount - 1) min 0
                            notifyItemChanged(pos)
                        }
                    } else {
                        MemeItMemes.likeComment(comment.id!!).call {
                            comment.likeCount++
                            if (comment.isDislikedByMe) {
                                comment.isDislikedByMe = false
                                comment.dislikeCount = (comment.dislikeCount - 1) min 0
                            }
                            comment.isLikedByMe = true
                            notifyItemChanged(pos)
                        }
                    }
                }
                R.id.dislike_comment, R.id.dislike -> {
                    if (comment.isDislikedByMe) {
                        MemeItMemes.removeDislikeComment(comment.id!!).call {
                            comment.dislikeCount = (comment.dislikeCount - 1) min 0
                            dislikeCount.text = comment.dislikeCount.toString()
                        }
                    } else {
                        MemeItMemes.dislikeComment(comment.id!!).call {
                            comment.dislikeCount++
                            if (comment.isLikedByMe) {
                                comment.isLikedByMe = false
                                comment.likeCount = (comment.likeCount - 1) min 0
                            }
                            comment.isDislikedByMe = true
                            notifyItemChanged(pos)
                        }
                    }
                }
                R.id.delete -> MaterialDialog.Builder(context)
                        .title("Delete comment?")
                        .positiveText("Yes")
                        .negativeText("No")
                        .onPositive { _, _ ->
                            MemeItMemes.deleteComment(comment.memeID!!, comment.id!!).call({
                                remove(comment)
                                notifyDataSetChanged()
                                itemView.snack("Deleted.")
                            }) {
                                itemView.snack("An error has occurred. Please try again.\n$it")
                            }
                        }.show()
                R.id.edit -> MaterialDialog.Builder(context)
                        .title("Edit comment")
                        .input("Comment", comment.comment, false, MaterialDialog.InputCallback { _, input ->
                            if (TextUtils.isEmpty(input))
                                return@InputCallback
                            val c = Comment(id = comment.id, comment = input.toString())
                            MemeItMemes.updateComment(c).call({
                                itemView.snack("Edited.")
                                comment.comment = input.toString()
                                notifyItemChanged(item_position)
                            }) {
                                itemView.snack("Editing has failed because of a problem. Please try again.\n$it" +
                                        "")

                            }
                        }).show()
            }
        }
    }
}
